### PR TITLE
update modify.py for rename_genes

### PR DESCRIPTION
**Pull request by maureencarey**
updated to permit modifications to a subset of gene ids and to avoid adding genes without GPRs

see below for example motivation:

```python
import cobra
from cobra import Model, Reaction, Metabolite
import os
print cobra.__version__
```

    0.8.2

```python
model = Model()
reaction = Reaction('reaction1')
reaction.name = 'first reaction'
reaction.lower_bound = 0.  # This is the default
reaction.upper_bound = 1000.  # This is the default
met1_c = Metabolite('met1_c', formula='C', name='metabolite 1', compartment='c')
met2_c = Metabolite('met2_c', formula='C', name='metabolite 2', compartment='c')
met3_c = Metabolite('met3_c', formula='C', name='metabolite 3', compartment='c')
met4_c = Metabolite('met4_c', formula='C', name='metabolite 4', compartment='c')
reaction.add_metabolites({met1_c: -1.0, met2_c: 1.0, met3_c: 0, met4_c: 0 })
reaction.gene_reaction_rule = '( old_gene1a and old_gene1b )'
model.add_reactions([reaction])

reaction = Reaction('reaction2')
reaction.name = 'second reaction'
reaction.lower_bound = 0.  # This is the default
reaction.upper_bound = 1000.  # This is the default
reaction.add_metabolites({met1_c: 0, met2_c: -1.0, met3_c: +1.0, met4_c: 0 })
reaction.gene_reaction_rule = '( old_gene2 )'
model.add_reactions([reaction])

reaction = Reaction('reaction3')
reaction.name = 'third reaction'
reaction.lower_bound = 0.  # This is the default
reaction.upper_bound = 1000.  # This is the default
reaction.add_metabolites({met1_c: 0, met2_c: 0, met3_c: -1.0, met4_c: +1.0 })
reaction.gene_reaction_rule = '( old_gene3 )'
model.add_reactions([reaction])

reaction = Reaction('ex1')
reaction.name = 'exchange reaction'
reaction.lower_bound = 0.  # This is the default
reaction.upper_bound = 1000.  # This is the default
reaction.add_metabolites({met1_c: +1.0, met2_c: 0, met3_c: 0, met4_c: 0 })
reaction.gene_reaction_rule = '( old_gene4 or old_gene5 )'
model.add_reactions([reaction])

reaction = Reaction('ex2')
reaction.name = 'exchange reaction'
reaction.lower_bound = 0.  # This is the default
reaction.upper_bound = 1000.  # This is the default
reaction.add_metabolites({met1_c: 0, met2_c: 0, met3_c: 0, met4_c: -1.0 })
model.add_reactions([reaction])

model.objective = 'reaction3'
print "objective reaction:"
print model.objective.expression

print ""
print "Reactions:"
for x in model.reactions:
    print "%s : %s" % (x.id, x.reaction)

print ""
print("Metabolites:")
for x in model.metabolites:
    print '%9s : %s' % (x.id, x.formula)

print ""
print "Genes:"
for x in model.genes:
    associated_ids = (i.id for i in x.reactions)
    print "%s is associated with reactions: %s" %(x.id, "{" + ", ".join(associated_ids) + "}")
```

    objective reaction:
    -1.0*reaction3_reverse_eb091 + 1.0*reaction3
    
    Reactions:
    reaction1 : met1_c --> met2_c
    reaction2 : met2_c --> met3_c
    reaction3 : met3_c --> met4_c
    ex1 :  --> met1_c
    ex2 : met4_c --> 
    
    Metabolites:
       met2_c : C
       met1_c : C
       met3_c : C
       met4_c : C
    
    Genes:
    old_gene1b is associated with reactions: {reaction1}
    old_gene1a is associated with reactions: {reaction1}
    old_gene2 is associated with reactions: {reaction2}
    old_gene3 is associated with reactions: {reaction3}
    old_gene5 is associated with reactions: {ex1}
    old_gene4 is associated with reactions: {ex1}

```python
dictionary = {'old_gene1a':'new_gene1',
              'old_gene1b':'new_gene1',
              'old_gene2':'new_gene2',
              'old_gene3':'new_gene3', 
              'old_gene4':'old_gene4', 
              'old_gene5':'new_gene5',
              'extra':'blahblahblah'}
```

```python
model_change = model.copy()
cobra.manipulation.modify.rename_genes(model_change, dictionary)
```

```python
print 'genes'
print [x.id for x in model_change.genes]
print ''
print 'GPRs'
print [x.gene_reaction_rule for x in model_change.reactions]
# note:
# [1] how blah blah blah was added even though it does not have a reaction association, 
# [2] old_gene_4 was deleted from the gene list even though it DOES have a reaction association, and
# [3] for the first reaction, the GPR is 'gene1 or gene1'.
```

    genes
    ['new_gene1', 'new_gene2', 'new_gene3', 'new_gene5', 'blahblahblah']
    
    GPRs
    ['new_gene1 and new_gene1', 'new_gene2', 'new_gene3', 'old_gene4 or new_gene5', '']

```python
model2 = model_change.copy()
# errors because old_gene4 is in in GPRs but not in gene list
```

    ---------------------------------------------------------------------------

    KeyError                                  Traceback (most recent call last)

    <ipython-input-6-f7cd6c3b56c0> in <module>()
    ----> 1 model2 = model_change.copy()
          2 # errors because old_gene4 is in in GPRs but not in gene list

    /Users/Mo/.virtualenvs/cobra_py2/lib/python2.7/site-packages/cobra/core/model.pyc in copy(self)
        298                 new_met._reaction.add(new_reaction)
        299             for gene in reaction._genes:
    --> 300                 new_gene = new.genes.get_by_id(gene.id)
        301                 new_reaction._genes.add(new_gene)
        302                 new_gene._reaction.add(new_reaction)

    /Users/Mo/.virtualenvs/cobra_py2/lib/python2.7/site-packages/cobra/core/dictlist.pyc in get_by_id(self, id)
         52     def get_by_id(self, id):
         53         """return the element with a matching id"""
    ---> 54         return list.__getitem__(self, self._dict[id])
         55 
         56     def list_attr(self, attribute):

    KeyError: 'old_gene4'

```python
rename_dict = dictionary.copy()
cobra_model = model.copy()

from six import iteritems
from ast import NodeTransformer
from cobra.core import Gene, Metabolite, Reaction
from cobra.core.gene import ast2str
from cobra.manipulation.delete import get_compiled_gene_reaction_rules

recompute_reactions = set()  # need to recomptue related genes
remove_genes = []
for old_name, new_name in iteritems(rename_dict):
    # undefined if there a value matches a different key
    # because dict is unordered
    try:
        gene_index = cobra_model.genes.index(old_name)
    except ValueError:
        gene_index = None
    old_gene_present = gene_index is not None
    new_gene_present = new_name in cobra_model.genes
    if old_gene_present and new_gene_present:
        old_gene = cobra_model.genes.get_by_id(old_name)
        if old_gene != cobra_model.genes.get_by_id(new_name): # Added in case not renaming some
            remove_genes.append(old_gene)
            recompute_reactions.update(old_gene._reaction)
    elif old_gene_present and not new_gene_present:
        # rename old gene to new gene
#         old_gene = cobra_model.genes.get_by_id(old_name) #added
        gene = cobra_model.genes[gene_index]
        # trick DictList into updating index
        cobra_model.genes._dict.pop(gene.id)  # ugh
        gene.id = new_name
        cobra_model.genes[gene_index] = gene
#         recompute_reactions.update(old_gene._reaction) # ADDED
    elif not old_gene_present and new_gene_present:
        pass # already fixed
    else:  # not old gene_present and not new_gene_present
        # the new gene's _model will be set by repair
        # cobra_model.genes.append(Gene(new_name)) # Removed, otherwise, adds genes that are unassigned to reactions
        pass

cobra_model.repair()

class Renamer(NodeTransformer):
    def visit_Name(self, node):
        node.id = rename_dict.get(node.id, node.id)
        return node

gene_renamer = Renamer()
for rxn, rule in iteritems(get_compiled_gene_reaction_rules(cobra_model)):
    if rule is not None:
        rxn._gene_reaction_rule = ast2str(gene_renamer.visit(rule))

for rxn in recompute_reactions:
    rxn.gene_reaction_rule = rxn._gene_reaction_rule
for i in remove_genes:
    cobra_model.genes.remove(i)

```

```python
print 'genes'
print [x.id for x in cobra_model.genes]
print ''
print 'GPRs'
print [x.gene_reaction_rule for x in cobra_model.reactions]
# note problems [3] not fixed (GPR1 is gene1 or gene1)
```

    genes
    ['new_gene1', 'new_gene2', 'new_gene3', 'new_gene5', 'old_gene4']
    
    GPRs
    ['new_gene1 and new_gene1', 'new_gene2', 'new_gene3', 'old_gene4 or new_gene5', '']

### CODE REVIEW

File: cobra/manipulation/modify.py [hunk 1]
```diff
@@ -93,7 +94,8 @@ def rename_genes(cobra_model, rename_dict):
             pass
         else:  # not old gene_present and not new_gene_present
             # the new gene's _model will be set by repair
-            cobra_model.genes.append(Gene(new_name))
+            # cobra_model.genes.append(Gene(new_name)) # MAC Sept 2017 edit: Removed, otherwise, adds genes that are unassigned to reactions
```
Comment (cdiener) [inline cobra/manipulation/modify.py]: You can just delete the old line. Git keeps track of the history anyways :)

Comment (maureencarey) [reply cobra/manipulation/modify.py]: thanks! I'll make the test!

**Comment by maureencarey** [comment]
looks like I created some version issues, so I will open a new pull request with the final updates to rename_genes and associated test.

**Comment by Midnighter** [comment]
Unfortunately, there have been quite a few changes on our devel branch. In future, I recommend creating a feature branch. You can then update your own devel branch with cobrapy/devel and then rebase your feature branch onto the latest devel.

**Pull request closed without merging**